Julia crashes on musl libc systems

Julia is on the process of being ported to a musl libc system (Alpine Linux).
However, there is a bug about "illegal instructions" making it crash.

```
$ julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.5
 _/ |\__'_|_|_|\__'_|  |
|__/                   |  x86_64-alpine-linux-musl


signal (4): Illegal instruction
jl_unprotect_stack at /usr/bin/../lib/julia/libjulia.so (unknown line)
jl_unprotect_stack at /usr/bin/../lib/julia/libjulia.so (unknown line)
jl_switchto at /usr/bin/../lib/julia/libjulia.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
jl_apply_generic at /usr/bin/../lib/julia/libjulia.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
jl_apply_generic at /usr/bin/../lib/julia/libjulia.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
_init at /usr/lib/julia/sys.so (unknown line)
jl_apply_generic at /usr/bin/../lib/julia/libjulia.so (unknown line)
unknown function (ip: 0x7f4d6b945d67)
unknown function (ip: 0x7f4d6b9458bf)
__libc_start_main at /lib/ld-musl-x86_64.so.1 (unknown line)
Illegal instruction
```

When using `--precompiled=no`:

```
$ julia --precompiled=no
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.5
 _/ |\__'_|_|_|\__'_|  |
|__/                   |  x86_64-alpine-linux-musl


signal (4): Illegal instruction
jl_unprotect_stack at /usr/bin/../lib/julia/libjulia.so (unknown line)
jl_unprotect_stack at /usr/bin/../lib/julia/libjulia.so (unknown line)
jl_switchto at /usr/bin/../lib/julia/libjulia.so (unknown line)
yieldto at ./task.jl:71
wait at ./task.jl:371
Illegal instruction
```

If `/usr/lib/julia/sys.so` is deleted, then:

```
$ julia
Illegal option -p
ERROR: system image file "/usr/bin/../lib/julia/sys.so" not found
```

It has also been tried to run it on a vanilla kernel (Alpine by default has a grsec kernel) but the exact same errors were observed.

Recompiling OpenBLAS without `DYNAMIC_ARCH` or setting `JULIA_CPU_TARGET` didn't make any difference either.

For the Julia developers that don't run Alpine Linux, here's a Dockerfile to help you reproduce the issue:

```
FROM alpine
RUN echo "http://dl-cdn.alpinelinux.org/alpine/v3.4/main" > /etc/apk/repositories
RUN echo "http://dl-cdn.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories
RUN echo "https://alpine.geeknet.cz/packages/edge/testing" >> /etc/apk/repositories
RUN wget -O "/etc/apk/keys/jakub@jirutka.cz-56d0d9fd.rsa.pub"
RUN https://alpine.geeknet.cz/keys/jakub%40jirutka.cz-56d0d9fd.rsa.pub
RUN apk update
RUN apk add julia julia-dev
RUN julia
```

Thanks.

/cc @tkelman @jirutka

> FWIW, I don't think this function is being called on master and you should probably have a try of master first.

I’ve tried master and it even doesn’t compile…

```
src/debuginfo.cpp:858:76: error: 'RTLD_DL_LINKMAP' was not declared in this scope
     dladdr_success = dladdr1((void*)pointer, &dlinfo, (void**)&extra_info, RTLD_DL_LINKMAP) != 0;
                                                                            ^
src/debuginfo.cpp:858:91: error: 'dladdr1' was not declared in this scope
     dladdr_success = dladdr1((void*)pointer, &dlinfo, (void**)&extra_info, RTLD_DL_LINKMAP) != 0;
                                                                                           ^
make[1]: *** [Makefile:116: debuginfo.o] Error 1
```

Where does this constant come from?

--- a/src/gc.c
+++ b/src/gc.c
@@ -57,10 +57,7 @@ arraylist_t to_finalize;
 
 NOINLINE uintptr_t gc_get_stack_ptr(void)
 {
-    void *dummy = NULL;
-    // The mask is to suppress the compiler warning about returning
-    // address of local variable
-    return (uintptr_t)&dummy & ~(uintptr_t)15;
+    return (uintptr_t)jl_get_frame_addr();
 }
 
 #define should_timeout() 0

--- a/src/julia_internal.h
+++ b/src/julia_internal.h
@@ -651,6 +651,18 @@ jl_value_t *jl_lookup_match(jl_value_t *a, jl_value_t *b, jl_svec_t **penv, jl_s
 
 unsigned jl_special_vector_alignment(size_t nfields, jl_value_t *field_type);
 
+STATIC_INLINE void *jl_get_frame_addr(void)
+{
+#ifdef __GNUC__
+    return __builtin_frame_address(0);
+#else
+    void *dummy = NULL;
+    // The mask is to suppress the compiler warning about returning
+    // address of local variable
+    return (void*)((uintptr_t)&dummy & ~(uintptr_t)15);
+#endif
+}
+
 #ifdef __cplusplus
 }
 #endif

--- a/src/task.c
+++ b/src/task.c
@@ -144,8 +144,8 @@ static void NOINLINE save_stack(jl_tls_states_t *ptls, jl_task_t *t)
 {
     if (t->state == done_sym || t->state == failed_sym)
         return;
-    volatile char *_x;
-    size_t nb = (char*)ptls->stackbase - (char*)&_x;
+    char *frame_addr = (char*)jl_get_frame_addr();
+    size_t nb = (char*)ptls->stackbase - frame_addr;
     char *buf;
     if (t->stkbuf == NULL || t->bufsz < nb) {
         buf = (char*)allocb(nb);
@@ -156,7 +156,7 @@ static void NOINLINE save_stack(jl_tls_states_t *ptls, jl_task_t *t)
         buf = (char*)t->stkbuf;
     }
     t->ssize = nb;
-    memcpy(buf, (char*)&_x, nb);
+    memcpy(buf, frame_addr, nb);
     // this task's stack could have been modified after
     // it was marked by an incremental collection
     // move the barrier back instead of walking it again here

